packages/xo-server/src/api/vm.mjs

```diff
@@ -1046,27 +1042,37 @@ export const stop = defer(async function ($defer, { vm, force, bypassBlockedOper
     )
   }
 
-  // Hard shutdown
-  if (force) {
-    return xapi.shutdownVm(vm._xapiRef, { hard: true })
-  }
-
   // Clean shutdown
-  try {
-    await xapi.shutdownVm(vm._xapiRef)
-  } catch (error) {
-    const { code } = error
-    if (code === 'VM_MISSING_PV_DRIVERS' || code === 'VM_LACKS_FEATURE_SHUTDOWN') {
-      throw invalidParameters('clean shutdown requires PV drivers')
+  let timedOut = false
+  if (!force || typeof force === 'number') {
+    const shutdownPromise = xapi.shutdownVm(vm._xapiRef).catch(error => {
+      const { code } = error
+      if (code === 'VM_MISSING_PV_DRIVERS' || code === 'VM_LACKS_FEATURE_SHUTDOWN') {
+        throw invalidParameters('clean shutdown requires PV drivers')
+      }
+      throw error
+    })
+
+    if (typeof force === 'number') {
+      await timeout.call(shutdownPromise, force).catch(error => {
+        if (error instanceof TimeoutError) {
+          timedOut = true
+        } else throw error
+      })
+    } else {
+      await shutdownPromise
     }
+  }
 
-    throw error
+  // Hard shutdown
+  if (force === true || timedOut) {
+    return xapi.shutdownVm(vm._xapiRef, { hard: true })
   }
 })
 
 stop.params = {
   id: { type: 'string' },
-  force: { type: 'boolean', optional: true },
+  force: { type: ['boolean', 'number'], optional: true },
```

Comment:
What do you think of introduce a new param for this use case, something like `forceShutdownDelay`?

Would that answer the issues you discussed with @MathieuRA?

Reply:
No, `force=true` immediately hard shutdown the VM.

`force=false forceShutdownDelay=60e3` attempt to clean shutdown the VM but then hard shutdown it after one minute.

Reply:
Ok, so in case force = true, we ignore `forceShutdownDelay` value ?

Reply:
Yes